src/tools/r2rdump/R2RReader.cs

```diff
@@ -65,25 +76,174 @@ public unsafe R2RReader(string filename)
                 Machine = peReader.PEHeaders.CoffHeader.Machine;
                 ImageBase = peReader.PEHeaders.PEHeader.ImageBase;
 
-                SectionHeader textSection;
-                int nSections = peReader.PEHeaders.CoffHeader.NumberOfSections;
-                for (int i = 0; i < nSections; i++)
-                {
-                    SectionHeader section = peReader.PEHeaders.SectionHeaders[i];
-                    if (section.Name.Equals(".text"))
-                    {
-                        textSection = section;
-                    }
-                }
-
+                // initialize the R2RHeader
                 DirectoryEntry r2rHeaderDirectory = peReader.PEHeaders.CorHeader.ManagedNativeHeaderDirectory;
-                int r2rHeaderOffset = r2rHeaderDirectory.RelativeVirtualAddress - textSection.VirtualAddress + textSection.PointerToRawData;
-                R2RHeader = new R2RHeader(_image, (uint)r2rHeaderDirectory.RelativeVirtualAddress, r2rHeaderOffset);
+                int r2rHeaderOffset = GetOffset(r2rHeaderDirectory.RelativeVirtualAddress);
+                R2RHeader = new R2RHeader(_image, r2rHeaderDirectory.RelativeVirtualAddress, r2rHeaderOffset);
                 if (r2rHeaderDirectory.Size != R2RHeader.Size)
                 {
                     throw new System.BadImageFormatException("The calculated size of the R2RHeader doesn't match the size saved in the ManagedNativeHeaderDirectory");
                 }
+
+                if (peReader.HasMetadata)
+                {
+                    var mdReader = peReader.GetMetadataReader();
+
+                    // get mapping from rid to MethodDefHandle
+                    int methodDefEntryPointsRVA = R2RHeader.Sections[R2RSection.SectionType.READYTORUN_SECTION_METHODDEF_ENTRYPOINTS].RelativeVirtualAddress;
+                    int methodDefEntryPointsOffset = GetOffset(methodDefEntryPointsRVA);
+                    Dictionary<uint, MethodDefinitionHandle> ridToMethodDefHandle = new Dictionary<uint, MethodDefinitionHandle>();
```

Comment:
This dictionary should not be needed. You can just call `MetadataTokens.MethodDefinitionHandle(rid)` to convert rid to MethodHandle. (This is very lightweight method, cheaper than dictionary lookup.)